### ISSUE TITLE
Can we unify Relation.Binary.Morphism.Definitions and Function.Definitions?

### ISSUE THREAD

**Issue by Taneb** [agda/agda-stdlib#2318]
`Relation.Binary.Morphism.Definitions` has as module parameters the domain and codomain of the functions over which the definitions are made, whereas `Function.Definitions` has them in a `variable` block. The former module only contains `Homomorphic₂`, which is exactly the same as the latter's `Congruent`, and a re-export of `Morphism` from `Function.Core`, which I think is otherwise unused in the library (and is just a synonym for the function arrow).

Can we deprecate `Relation.Morphism.Definitions` in favour of `Function.Definitions`?

**Comment by jamesmckinna** [comment]
"Can"? or "Should"?

It seems as though (modulo dependencies/dependency cycles? not sure about this...) we can.

Not obvious to me that we should... partly because of the ergonomics of the whole `Algebra` hierarchy... that eventually rests on `Relation.Binary` (and `IsRelHomomorphism` specifically), and swapping out a `Relation.Binary` component in favour of a `Function` one, even if semantically equivalent/justified, seems like a rupture (however small) in the design hierarchy? Maybe this is a nonsense argument, but the "what are all the things I need to import to get task X done?" question (allied with the lack of suitable search tools) means that I like not having to look 'outside the box' for functionality if I can avoid it... but this is off the top of my head, without a serious analysis of the situation.

**Comment by MatthewDaggitt** [comment]
I agree with @jamesmckinna. We absolutely can, but I don't think we should because it would be violating the standard design of hierarchies in the library. Obviously if we could do it for every hierarchy in the library that would be a different question....

**Comment by jamesmckinna** [comment]
... And the above remarks apply also to `Algebra.Consequences`, where it would (perhaps?) be better ergonomically not to have to reimport `Function.Definitions` instantiated twice with `_≈_`...

**Comment by jamesmckinna** [comment]
See also #2581 for yet another variation on this theme...

**Comment by JacquesCarette** [comment]
About your two different `Foo` modules above: my impression was that, at the usage site, there are the same. So if they are not, that should warrant an explanation in the official documentation.

**Comment by jamesmckinna** [comment]
I'll refine the example to explain how I think they *could* be different at usage sites... later

**Comment by jamesmckinna** [comment]
Looping back to @Taneb 's OP, and mine and @MatthewDaggitt 's comments, there's a separate problem with 'too much' DRY-delegation of one hierarchy to another: the cognitive load on the user/reader in needing definitional equality to be 'transparent' across modules... it  is, semantically, of course, but not 'visually'. Hence the recent discussion/critique of #2566 / #2581

**Comment by JacquesCarette** [comment]
Or more succinctly: mentally expanding definitions has non-zero cognitive load.

**Comment by Taneb** [comment]
Closing this, as I think it's been answered "we shouldn't do this even if we technically can" and the discussion has gone _way_ further than I expected it would when I opened it.
